### DISCUSSION TITLE
I can not run Ws2812B LED by sample of nanoFramework.IoT.Device

### DISCUSSION THREAD

**Discussion by huster-songtao** [nanoframework/Home#1386] [Q&A]
### Location

GitHub README

### Description

[https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/Ws28xx/README.md](url)

`Configuration.SetPinFunction(22, DeviceFunction.ADC1_CH10);`

It throw System.ArgumentException.
![20231020145936](https://github.com/nanoframework/Home/assets/94954133/0446d85c-c98b-4b62-8e4f-044637276b70)

### Screenshots

![20231020145936](https://github.com/nanoframework/Home/assets/94954133/75c5740f-2d55-434b-9482-cdb810d5f633)

**Comment by josesimoes** [comment]
As mentioned in the issues instructions, we're not using these to answer questions.
Moving this to discussions.

**Reply by huster-songtao** [reply]
![image](https://github.com/nanoframework/Home/assets/94954133/b7780c14-eaa3-4041-b28e-6026ef622e6e)

LED   -> ESP32-DevKitC V4（+ ESP32-WROOM-32D）
GNG -> GNG
IN -> 23

My ESP32 development board is wired like this:
![20231020163637](https://github.com/nanoframework/Home/assets/94954133/63b94e37-ece1-4191-bc0f-ba3e07e0acde)

It work well using Iot.Device.Ws28xx.Esp32.Ws2808.
```
            // GRB
            Light(255, 0, 0, 23, 0, 0, 2000); // 255, 0, 0 Green

            Light(0, 255, 0, 23, 0, 0, 2000); // 0, 0, 255 Red

            Light(0, 0, 255, 23, 0, 0, 2000); // 0, 0, 255 Blue

            static void Light(int red, int green, int blue, int gpioPin, int startPixel, int endPixel, int milliseconds = 1)
            {
                int pixelCount = endPixel + 1;
                var neo = new Iot.Device.Ws28xx.Esp32.Ws2808(gpioPin, pixelCount);
                var img = neo.Image;

                for (var index = startPixel; index <= endPixel; index++)
                {
                    img.SetPixel(index, 0, Color.FromArgb(red, green, blue));
                }
                neo.Update();
                Thread.Sleep(milliseconds);
            }
```

It work well on Iot.Device.Ws28xx.Esp32.Ws2808, but I want to run WS2818B on Iot.Device.Ws28xx using SPI.

```
// Configure the count of pixels
const int Count = 8;
Console.Clear();

// Must specify pin functions on ESP32

Configuration.SetPinFunction(23, DeviceFunction.SPI2_MOSI);
Configuration.SetPinFunction(19, DeviceFunction.SPI2_MISO);
Configuration.SetPinFunction(18, DeviceFunction.SPI2_CLOCK);
Configuration.SetPinFunction(22, DeviceFunction.ADC1_CH10);

```

The code above throw exception at 'Configuration.SetPinFunction(22, DeviceFunction.ADC1_CH10);'
I have no idea to resolve this exception.
I searched Bing and Google and found all the links or articles related to ESP32, SPI, and WS2812B, but couldn't find a solution.

**Reply by Ellerbach** [reply]
@huster-songtao do **not** set this one: Configuration.SetPinFunction(22, DeviceFunction.ADC1_CH10);

Also for performances reasons, you should not instanciate and deinstanciate this: var neo = new Iot.Device.Ws28xx.Esp32.Ws2808(gpioPin, pixelCount);
create it for the full project especially as you are using the same pin.

**Reply by huster-songtao** [reply]
do not set this one: Configuration.SetPinFunction(22, DeviceFunction.ADC1_CH10);
The problem is the same, the LED light strip cannot be lit

**Comment by josesimoes** [comment]
This is unrelate with Ws2812B... the issue seems to be assigning GPIO22 to ADC CH10. 
Is my understanding correct?

**Comment by huster-songtao** [comment]
[https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/Ws28xx/README.md](https://github.com/nanoframework/Home/discussions/url)
The code example I ran directly did not add any additional code, which failed to run. However, I couldn't find a solution, 

so I ran another code example. The WS2812B light board can light up, and the breathing light effect can be achieved with less than 24 light beads. However, the breathing light effect with 30 light beads feels slightly stuck, and the experience is not very good. 
```
            static void Breathe(int red, int green, int blue, int gpioPin, int startPixel, int pixelCount, float minThreshold = 0.05f, float maxThreshold = 1f, int milliseconds = 1)
            {
                // Use Ws2812 or SK6812 instead if needed
                int endPixel = startPixel + pixelCount;
                var neo = new Ws2812c(gpioPin, pixelCount);

                //// RGB=0,233,255
                // 24bit对应的的GRB编码如下所示。发送颜色顺序为GRB，字节的高位在前。
                //var rgb = color.Split(',');
                //if (rgb.Length != 3)
                //    return;

                var ts = 0;
                var add = false; // 从1递减，由亮变暗
                var rate = 1f;
                var step = 0.05f;
                var img = neo.Image;

                for (; ; )
                {
                    for (; ; )
                    {
                        if (rate > maxThreshold)
                        {
                            rate = maxThreshold;
                        }
                        if (rate < minThreshold)
                        {
                            rate = minThreshold;
                        }

                        var r = (int)(red * rate);
                        var g = (int)(green * rate);
                        var b = (int)(blue * rate);

                        for (var index = startPixel; index < endPixel; index++)
                        {
                            img.SetPixel(index, 0, Color.FromArgb(r, g, b));

                            //img.SetPixel(j, 0, Wheel((i + j) & 255));
                            //controller.SetHSVColor(index, 185, 100, v);

                            //img.SetPixel(index, 0, Color.FromArgb(r, g, b));
                        }
                        neo.Update(); // startPixel, pixelCount
                        Thread.Sleep(milliseconds);

                        if (add)
                        {
                            rate += step; 
                        }
                        else
                        {
                            rate -= step; 
                        }

                        if (rate > maxThreshold || rate < minThreshold)
                        {
                            add = !add;
                            break;
                        }
                    }

                    if (ts > 10) break;
                    ts++;
                }
            }
```

I directly provided a color array to reduce the computational complexity of the breathing effect, The effect still feels slightly stuck, but it is limited by the Dirver itself and there is not a good optimization approach. 
```
            var colors = new ArrayList {
                new int[3] { 0, 233, 255 },
                new int[3] { 0, 221, 242 },
                new int[3] { 0, 198, 217 },
                new int[3] { 0, 168, 184 },
                new int[3] { 0, 134, 147 },
                new int[3] { 0, 100, 110 },
                new int[3] { 0, 69, 76 },
                new int[3] { 0, 44, 49 },
                new int[3] { 0, 26, 29 },
                new int[3] { 0, 14, 15 },
                new int[3] { 0, 6, 7 },
                new int[3] { 0, 2, 3 }//,
                //new int[3] { 0, 0, 1 },
                //new int[3] { 0, 0, 0 }
            };

            while (true)
            {
                //Breathe(0, 233, 255, 23, 0, 300);
                Breathe2(23, colors, 0, 7);
            }

            static void Breathe2(int gpioPin, ArrayList colors, int startPixel, int pixelCount)
            {
                int steps = colors.Count;
                int endPixel = startPixel + pixelCount;
                var neo = new Ws2812c(gpioPin, pixelCount);
                var img = neo.Image;
                //int milliseconds = 10;
                 
                for (var index = 0; index < steps; ++index)
                {
                    var color = colors[index] as int[];
                    for (var pixel = startPixel; pixel < endPixel; ++pixel)
                    {
                        img.SetPixel(pixel, 0, Color.FromArgb(color[0], color[1], color[2]));
                    }
                    neo.Update();
                    //Thread.Sleep(milliseconds);
                }
                for (var index = steps - 1; index >= 0; --index)
                {
                    var color = colors[index] as int[];
                    for (var pixel = startPixel; pixel < endPixel; ++pixel)
                    {
                        img.SetPixel(pixel, 0, Color.FromArgb(color[0], color[1], color[2]));
                    }
                    neo.Update();
                    //Thread.Sleep(milliseconds);
                }
            }
```

The ESP32 is unable to light up 300 lights by nanoframework, which means it cannot light up a single light strip. This strip has a total of 60 light beads per meter, and a total of 5 meters per strip. An exception of outofmemory was throwed.

**Comment by huster-songtao** [comment]
https://www.bilibili.com/video/BV1j34y1d7kr/?spm_id_from=333.851.dynamic.content.click&vd_source=5ba4a29b69cce9fd14c9d0068672dc97
Video of ESP32+WS2812B Realizing Breath Lamp Effect
5 meters in length, 60 lamp beads per meter, powered by a 5V/3A charger, without using a battery powered solution.
The effect looks very good. Change the color to RGB=0,233,255, Implementing a breathing light effect in one color indicates that the device is working, the effect should be suitable for me to use.

I am very familiar with desktop application development and web application development in C #, but I am a bit inexperienced in developing ESP32 and light strips using nanoframework for the first time. I think the nanoframework and system have consumed too much memory, causing an exception of insufficient memory when 300 light bulbs are lit.

I wanted to make some contributions and made some attempts, but in the end, they all failed.

1. I downloaded the driver code Iot.Device.Ws28xx.Esp32, and the light strip traversed each light bead, then gave the color value. I should change this to from the specified light bead to a certain light bead at the back, but the running result was not very ideal.
\nanoFramework.IoT.Device\devices\Ws28xx.Esp32\Ws28xx.cs line 65
```
        /// <summary>
        /// Sends backing image to the LED driver.
        /// </summary>
        /// <param name="startPixel">Start Pixel. Which lamp bead should I start from</param>
        /// <param name="pixelCount">Pixel Count. How many lamp beads are ready to control</param>
        public void Update(int startPixel, int pixelCount)
        {
            int start = startPixel * 3;
            int end = pixelCount * 3;
            if (end > Image.Data.Length)
            {
                end = Image.Data.Length;
            }

            var transmitterSettings = new TransmitChannelSettings(pinNumber: GpioPin)
            {
                EnableCarrierWave = false,
                IdleLevel = false,

                // this value for the clock divider considers a clock source of 80MHz which is what we have fixed in native
                ClockDivider = this.ClockDivider,
            };
            var transmitter = new TransmitterChannel(transmitterSettings);

            for (int i = start; i < end; i++)
            {
                SerializeColor(Image.Data[i], transmitter);
            }

            transmitter.AddCommand(ResetCommand);
            transmitter.Send(true);
            transmitter.Dispose();
        }
```

I don't remember the effect of lighting upLight(255, 0, 0, 23, 0, 0, 2000);(), but the effect of the breathing light is very poor.
           ```
 //Light(255, 0, 0, 23, 0, 0, 2000); // 255, 0, 0 Red
            //Light(0, 255, 0, 23, 0, 0, 2000); // 0, 0, 255 Green
            //Light(0, 0, 255, 23, 0, 0, 2000); // 0, 0, 255 Blue
            static void Light(int red, int green, int blue, int gpioPin, int startPixel, int endPixel, int milliseconds = 1)
            {
                // 使用SPI驱动 Ws2812B 灯
                int pixelCount = endPixel + 1;
                var neo = new Ws2812c(gpioPin, 16);
                var img = neo.Image;

                for (var index = startPixel; index < pixelCount; index++)
                {
                    img.SetPixel(index, 0, Color.FromArgb(red, green, blue));

                }
                neo.Update(startPixel, pixelCount);

                Thread.Sleep(milliseconds);
            }
```

2. The lamp beads cannot be powered separately and can only be connected to the GND and 5V of the development board, which is a significant limitation. I connected the signal wire to pin 23 of ESP32, and connected the 5V and ground points to a dedicated power module, resulting in a dismal display. 
The power supply of the Ws2812B light strip must go through ESP32, which is very unreasonable. In many projects, the power supply of the light strip is entirely supplied by a dedicated power source, as there is a voltage drop issue with the light strip.
